component/src/action_handler/actions/spend.rs

```diff
@@ -44,9 +47,14 @@ impl ActionHandler for Spend {
         state.check_nullifier_unspent(spent_nullifier).await
     }
 
-    #[instrument(name = "spend", skip(self, _state))]
-    async fn execute(&self, _state: &mut StateTransaction) -> Result<()> {
-        // Handled in [`crate::action_handler::transaction::Transaction`].
+    #[instrument(name = "spend", skip(self, state))]
+    async fn execute(&self, state: &mut StateTransaction) -> Result<()> {
+        let source = state.object_get("source").cloned().unwrap_or_default();
+
+        state.spend_nullifier(self.body.nullifier, source).await;
+        // TODO: why do we manage event emission separately up at the top level
+        // instead of integrated into state machine?
+        state.record(shielded_pool::event::spend(self.body.nullifier));
```

Comment:
I think this is a holdover from before we combined the state system and the events system; it seems like it would be better to fold the event handling into the component logic.